exporter/exporterhelper/queued_retry_inmemory.go

```diff
@@ -71,49 +75,142 @@ func (qCfg *QueueSettings) Validate() error {
+	errMultipleStorageClients = errors.New("multiple storage extensions found while default extension expected")
+)
+
 type queuedRetrySender struct {
-	fullName        string
-	cfg             QueueSettings
-	consumerSender  requestSender
-	queue           internal.ProducerConsumerQueue
-	retryStopCh     chan struct{}
-	traceAttributes []attribute.KeyValue
-	logger          *zap.Logger
+	fullName           string
+	id                 config.ComponentID
+	signal             config.DataType
+	cfg                QueueSettings
+	consumerSender     requestSender
+	queue              internal.ProducerConsumerQueue
+	queueStartFunc     internal.PersistentQueueStartFunc
+	retryStopCh        chan struct{}
+	traceAttributes    []attribute.KeyValue
+	logger             *zap.Logger
+	requeuingEnabled   bool
+	requestUnmarshaler internal.RequestUnmarshaler
 }
 
-func newQueuedRetrySender(id config.ComponentID, _ config.DataType, qCfg QueueSettings, rCfg RetrySettings, _ internal.RequestUnmarshaler, nextSender requestSender, logger *zap.Logger) *queuedRetrySender {
+func newQueuedRetrySender(id config.ComponentID, signal config.DataType, qCfg QueueSettings, rCfg RetrySettings, reqUnmarshaler internal.RequestUnmarshaler, nextSender requestSender, logger *zap.Logger) *queuedRetrySender {
 	retryStopCh := make(chan struct{})
 	sampledLogger := createSampledLogger(logger)
 	traceAttr := attribute.String(obsmetrics.ExporterKey, id.String())
-	return &queuedRetrySender{
-		fullName: id.String(),
-		cfg:      qCfg,
-		consumerSender: &retrySender{
-			traceAttribute:     traceAttr,
-			cfg:                rCfg,
-			nextSender:         nextSender,
-			stopCh:             retryStopCh,
-			logger:             sampledLogger,
-			onTemporaryFailure: onTemporaryFailure,
-		},
-		queue:           internal.NewBoundedMemoryQueue(qCfg.QueueSize, func(item interface{}) {}),
-		retryStopCh:     retryStopCh,
-		traceAttributes: []attribute.KeyValue{traceAttr},
-		logger:          sampledLogger,
+
+	qrs := &queuedRetrySender{
+		fullName:           id.String(),
+		id:                 id,
+		signal:             signal,
+		cfg:                qCfg,
+		retryStopCh:        retryStopCh,
+		traceAttributes:    []attribute.KeyValue{traceAttr},
+		logger:             sampledLogger,
+		requestUnmarshaler: reqUnmarshaler,
+	}
+
+	qrs.consumerSender = &retrySender{
+		traceAttribute: traceAttr,
+		cfg:            rCfg,
+		nextSender:     nextSender,
+		stopCh:         retryStopCh,
+		logger:         sampledLogger,
+		// Following three functions actually depend on queuedRetrySender
+		onTemporaryFailure: qrs.onTemporaryFailure,
+	}
+
+	if qCfg.StorageEnabled || qCfg.StorageID != nil {
+		qrs.queue, qrs.queueStartFunc = internal.NewPersistentQueue(qrs.fullName, qrs.signal, qrs.cfg.QueueSize, qrs.logger, qrs.requestUnmarshaler)
+		// TODO: following can be further exposed as a config param rather than relying on a type of queue
+		qrs.requeuingEnabled = true
```

Comment:
Looks like this is a new functionality that didn't exist before. Can we remove this from this PR and add in a subsequent PR? This PR should be only about enabling Persistent Queue by default.

Reply:
It used to be enabled here: https://github.com/open-telemetry/opentelemetry-collector/blob/e8dd9536f5e53228606c325f12d4bfc4b4293acc/exporter/exporterhelper/queued_retry_experimental.go#L170 so this actually preserves existing behaviour.